File: inst/examples/e2e-test/R/deepMixing.R
deepMixingUI <- function(id) {
  ns <- NS(id)
  tagList(
    h1("Deep mixing of Shiny and react"),
    p(),
    div(
      h3("Fluent components inside Shiny tabset partially work"),
      tabsetPanel(
        id = ns("tabs1"),
        type = "tabs",
        tabPanel("Other panel", Label("Other")),
        tabPanel("JSX",
          Stack(
            horizontal = TRUE, 
            tokens = list(childrenGap = 20),
            Label("I'm a Label", `data-cy` = ns("label")),
            PrimaryButton(text = "I'm a Button", `data-cy` = ns("primaryButton"))
          )
        )
      ),
      h3("JSX inside uiOutput inside JSX"),
      uiOutput(ns("layer1"))
    ),
    PrimaryButton.shinyInput(ns("addPoint"), text = "Add point", `data-cy` = ns("addPoint"))
  )
}

deepMixingServer <- function(id) {
  moduleServer(id, function(input, output, session) {
    ns <- session$ns
    output$layer1 <- renderUI({
      Pivot(
        `data-cy` = ns("pivot"),
        PivotItem(headerText = "Tab 1", `data-cy` = "x", Label("Hello 1")),
        PivotItem(headerText = "Leaflet inside a Pivot", leafletOutput(ns("map"))),
        PivotItem(headerText = "Another level of nesting", uiOutput(ns("layer2")))
      )
    })
    
    output$layer2 <- renderUI({
      div(
        h3("Leaflet in Pivot in uiOutput in Pivot in uiOutput in JSX"),
        Pivot(
          `data-cy` = ns("nestedPivot"),
          PivotItem(headerText = "Tab 1", Label("Hello 2")),
          PivotItem(headerText = "Leaflet inside a Pivot", leafletOutput(ns("map")))
        )
      )
    })
    
    points <- eventReactive(input$addPoint, {
      n <- as.numeric(input$addPoint)
      cbind(rnorm(n) * 2 + 13, rnorm(n) + 48)
    }, ignoreNULL = FALSE, ignoreInit = FALSE)
    
    output$map <- renderLeaflet({
      leaflet() %>% addTiles() %>% addMarkers(data = points())
    })
  })
}

File: js/cypress/integration/e2e-test/deepMixing.js
describe('Fluent components inside Shiny tabset', () => {
  before(() => {
    cy.visit('/');
    cy.get('#deepMixing-tabs1').within(() => {
      cy.get('a[data-value=JSX]').click();
    });
  });

  it('Label is visible', () => {
    cy.get('[data-cy=deepMixing-label').should('be.visible');
  });

  it('PrimaryButton is visible', () => {
    cy.get('[data-cy=deepMixing-primaryButton').should('be.visible');
  });
});

describe('JSX inside uiOutput inside JSX', () => {
  before(() => {
    cy.get('[data-cy=deepMixing-pivot] > .ms-Pivot > button[name="Leaflet inside a Pivot"]').click();
  });

  it('Leaflet inside Pivot is visible', () => {
    cy.get('[data-cy=deepMixing-pivot] > div[role=tabpanel]').within(() => {
      cy.get('#deepMixing-map').should('be.visible');
    });
  });
});

describe('Leaflet in Pivot in uiOutput in Pivot in uiOutput in JSX', () => {
  before(() => {
    cy.get('[data-cy=deepMixing-pivot] > .ms-Pivot > button[name="Another level of nesting"]').click();
    cy.get('[data-cy=deepMixing-nestedPivot] > .ms-Pivot > button[name="Leaflet inside a Pivot"]').click();
  });

  it('Leaflet inside nested Pivot is visible', () => {
    cy.get('[data-cy=deepMixing-nestedPivot] > div[role=tabpanel]').within(() => {
      cy.get('#deepMixing-map').should('be.visible');
    });
  });
});
